packages/tre/src/plugins/r2/gateway.ts

```diff
@@ -177,22 +359,298 @@ function sqlStmts(db: TypedDatabase) {
+          if (partSize === undefined) partSize = part.size;
+          if (
+            part.size < R2Gateway._MIN_MULTIPART_PART_SIZE ||
+            part.size !== partSize
+          ) {
+            throw new BadUpload();
+          }
+        }
+        //    ...and the last part is not greater than all others
+        //    (if part size is defined, we must have at least one part)
+        if (partSize !== undefined && parts[parts.length - 1].size > partSize) {
+          throw new BadUpload();
+        }
+
+        // 5. Get existing upload if any, and delete previous multipart parts
+        const oldBlobIds: string[] = [];
+        const existingRow = stmtGetPreviousByKey.get({ key });
+        const maybeOldBlobId = existingRow?.blob_id;
+        if (maybeOldBlobId === null) {
+          // If blob_id is null, this was a multipart object, so delete all
+          // multipart parts
+          const partRows = stmtDeletePartsByKey.all({ object_key: key });
+          for (const partRow of partRows) oldBlobIds.push(partRow.blob_id);
+        } else if (maybeOldBlobId !== undefined) {
+          oldBlobIds.push(maybeOldBlobId);
+        }
+
+        // 6. Write object to the database, and link parts with object
+        const totalSize = parts.reduce((acc, { size }) => acc + size, 0);
+        const etag = generateMultipartEtag(
+          parts.map(({ checksum_md5 }) => checksum_md5)
+        );
+        const newRow: ObjectRow = {
+          key,
+          blob_id: null,
+          version: generateVersion(),
+          size: totalSize,
+          etag,
+          uploaded: Date.now(),
+          checksums: "{}",
+          http_metadata: uploadRow.http_metadata,
+          custom_metadata: uploadRow.custom_metadata,
+        };
+        stmtPut.run(newRow);
+        for (const part of parts) {
+          stmtLinkPart.run({
+            upload_id,
+            part_number: part.part_number,
+            object_key: key,
+          });
+        }
+
+        // 7. Delete unlinked, unused parts
+        const partRows = stmtDeleteUnlinkedPartsByUploadId.all({ upload_id });
+        for (const partRow of partRows) oldBlobIds.push(partRow.blob_id);
+
+        // 8. Mark the upload as completed
+        stmtUpdateUploadState.run({
+          upload_id,
+          state: MultipartUploadState.COMPLETED,
+        });
+
+        return { newRow, oldBlobIds };
+      }
+    ),
+    abortMultipartUpload: db.transaction((key: string, upload_id: string) => {
+      // 1. Make sure this multipart upload exists, ignoring finalised states
+      const uploadRow = stmtGetUploadState.get({ key, upload_id });
+      if (uploadRow === undefined) {
+        throw new InternalError();
+      } else if (uploadRow.state > MultipartUploadState.IN_PROGRESS) {
+        // If this upload has already been finalised, return here. `abort()` can
+        // be called multiple times, and on already `complete()`ed uploads. In
```

Comment:
I would have imagined that aborting a completed upload would be an error?
But if this is what R2 does then, heh?